epsie/proposals/adaptcov_normal.py

```diff
@@ -0,0 +1,213 @@
+
+
+    @property
+    def prior_widths(self):
+        """The width of the prior used for each parameter."""
+        return self._prior_widths
+
+    @prior_widths.setter
+    def prior_widths(self, prior_widths):
+        """Sets the prior widths, making sure that widths are provided for
+        each parameter in ``parameters``.
+
+        Also sets the deltas attribute.
+        """
+        try:
+            self._prior_widths = {p: abs(prior_widths[p])
+                                  for p in self.parameters}
+        except KeyError:
+            raise ValueError("must provide prior widths for every parameter")
+
+    @property
+    def adaptation_duration(self):
+        """The adaptation duration used."""
+        return self._adaptation_duration
+
+    @adaptation_duration.setter
+    def adaptation_duration(self, adaptation_duration):
+        """Sets the adaptation duration to the given value, making sure it is
+        larger than 1.
+        """
+        if adaptation_duration < 1:
+            raise ValueError("adaptation duration must be >= 1")
+        self._adaptation_duration = adaptation_duration
+
+    @property
+    def start_iteration(self):
+        """The iteration that the adaption begins."""
+        return self._start_iteration
+
+    @start_iteration.setter
+    def start_iteration(self, start_iteration):
+        """Sets the start iteration, making sure it is >= 1."""
+        if start_iteration < 1:
+            raise ValueError("start_iteration must be >= 1")
+        self._start_iteration = start_iteration
+
+    def update(self, chain):
+        """Updates the adaptation based on whether the last jump was accepted.
+        This prepares the proposal for the next jump.
+        """
+
+        if self.start_iteration < chain.iteration < self._adaptation_end:
+            # change this decay later. For now ok
+            decay = (chain.iteration - self.start_iteration)**(-0.2) - 0.1
+
+            # how to better unpack the last position?
+            names = chain.positions.dtype.names
+            lpos = chain.positions[-1]
+            newpt = np.array([lpos[p] for p in names])
+            # Update the first and second moments
+            new_mean = self._mean +  decay * (newpt - self._mean)
+            d = (newpt - self._mean).reshape(-1, 1)
+            new_cov = self._cov + decay * (np.matmul(d, d.T) - self._cov)
+
+            # this is to ensure that the logpdf method does not throw an error
+            # sometimes it returns SIngularMatrix even if it can sample a rvs
+            # but the logpdf method fails. ANyway there has to be some check
+            # for singular matrices
+            try:
+                __ = stats.multivariate_normal.logpdf(newpt, mean=new_mean,
+                                                    cov=new_cov)
+                self._mean = new_mean
+                self._cov = self._r**2 * new_cov
+            except:
+                pass
```

Comment:
Hmm... singular matrices are a problem. What you're doing here makes sense (although it's better practice to catch the specific error that gets raised, rather then doing a generic try/except, in case there is another error that you really do want to stop for). But what happens on the next iteration? Will you end up with the same covariance matrix, and so it never will adapt beyond this?

Reply:
I had trouble catching the specific exception.. it's np.linalg.LinAlgError. It must be related to numerical errors, i.e it works for generating samples but if you try to evaluate the logpdf it throws the error
The problem seems to be that the covariance is sometimes really small, resulting in huge inverse covariance